File: src/alpha/Physical.js
/* eslint-disable require-jsdoc, max-len, new-cap */
// Physical version 1.4.130828
// physical is an orientation and a position
// as well as rotation and movement

// TODO change the rotation movement speeds from x,y,z to forward, backward,left,right, etc
// // really front/back up/down lateral(surely we don't want crippled things )
// // // this also requires rethinking the movement; because moveforward/backward won't cancel anymore
// TODO: scaling
// TODO: tilt
// TODO: Children
// TODO: acceleration

import TestSuite from 'parsegraph-testsuite';

// -----------------------------------
// ------------ USAGE ----------------
// -----------------------------------

// local p = Physical();

// p:SetPosition(x,y,z);
// p:ChangePosition(x,y,z); // adds to current global position
// p:Rotate( angle, x, y, z ) // rotates at its current position, using p's x,y,z axes

// glMultMatrix( p() ) // applying the above

// -----------------------------------
// --------- BETTER USAGE ------------
// -----------------------------------
// speeds only apply to functions taking an elapsed parameter

// p:SetRotationSpeeds( x, y, z ) // radians / second per second per axis
// SetRotationSpeed( speed ) // sets all axes the same
// p:SetSpeeds(x,y,z) // ( movement speeds ) units / second per axis

// p:YawLeft( elapsed ) // p:YawRight( elapsed )
// p:PitchUp( elapsed ) // p:PitchDown( elapsed )
// p:RollLeft( elapsed ) // p:RollRight( elapsed )

// instantly update your global position when you call these
// p:WarpForward( distance ) // p:WarpBackward( distance )
// p:WarpLeft( distance ) // p:WarpRight( distance )
// p:WarpUp( distance ) // p:WarpDown( distance )

// velocity is applied whenever you call p:ApplyVelocity() or p:GetMatrix()
// velocity will adjust your current position by the velocity;
// p:SetVelocity(x,y,z);

// a simpler way to use velocity is to use these:
// it will be automatically calculated using our set speeds;

// p:MoveForward( elapsed ) // p:MoveBackward( elapsed )
// p:MoveLeft( elapsed ) // p:MoveRight( elapsed )
// p:MoveUp( elapsed ) // p:MoveDown( elapsed )

// XXX: for some reason I have to inverse quaterions for physical
// not for the camera. I do not understand why.

PHYSICAL_TRANSLATE_ROTATE_SCALE = 1;
PHYSICAL_SCALE_ROTATE_TRANSLATE = 2;
PHYSICAL_ROTATE_TRANSLATE_SCALE = 3;

export default function Physical(parent) {
  this.modelMode = PHYSICAL_TRANSLATE_ROTATE_SCALE;
  this.orientation = new Quaternion();
  this.position = new Vector();
  this.modelMatrix = new RMatrix4();
  this.viewMatrix = new RMatrix4();
  this.modelDirty = false; // whether or not the matrix needs to be updated;
  this.velocity = new Vector();
  this.rotationSpeed = new Vector(1, 1, 1);
  this.speed = new Vector(5, 5, 5);
  this.scale = new Vector(1, 1, 1);
  this.SetParent(parent);
}

Physical.prototype.toJSON = function() {
  return {
    position: this.position.toJSON(),
    orientation: this.orientation.toJSON(),
  };
};

// Register the test suite.
alphaPhysicalTests = new TestSuite('Physical');

alphaPhysicalTests.addTest('Physical', function(resultDom) {
  const surface = new GLWidget();
  const cam = new Camera(surface);
  const p = new Physical(cam);
});

// -----------------------------------
// ---------- Rotation ---------------
// -----------------------------------

Physical.prototype.SetOrientation = function(...args) {
  this.orientation.Set.apply(this.orientation, ...args);
  this.modelDirty = true;
};

/*
 * returns as Quaternion
 */
Physical.prototype.GetOrientation = function() {
  return this.orientation;
};

/*
 * in radians / second
 */
Physical.prototype.SetRotationSpeeds = function(...args) {
  this.rotationSpeed.Set.apply(this.rotationSpeed, ...args);
};
Physical.prototype.SetRotationSpeed =
  Physical.prototype.SetRotationSpeeds;

Physical.prototype.GetRotationSpeeds = function() {
  return this.rotationSpeed;
};

Physical.prototype.Rotate = function(angle, x, y, z) {
  // if you aren't rotating about an angle, then you aren't rotating
  if (angle == 0) {
    return;
  }
  const q = QuaternionFromAxisAndAngle(x, y, z, angle);
  this.orientation.Multiply(q);
  this.modelDirty = true;
};

Physical.prototype.RotateGlobal = function(angle, x, y, z) {
  // if you aren't rotating about an angle, then you aren't rotating
  if (angle == 0) {
    return;
  }
  const q = QuaternionFromAxisAndAngle(x, y, z, angle);
  this.orientation.Set(q.Multiply(this.orientation));
  this.modelDirty = true;
};

/*
 * these rotations take place at the speeds set by rotationSpeed
 */
Physical.prototype.YawLeft = function(elapsed) {
  const angle = elapsed * this.rotationSpeed[1];
  this.Rotate(angle, 0, 1, 0);
};

Physical.prototype.YawRight = function(elapsed) {
  const angle = elapsed * this.rotationSpeed[1];
  this.Rotate(-angle, 0, 1, 0);
};

Physical.prototype.PitchUp = function(elapsed) {
  const angle = elapsed * this.rotationSpeed[0];
  this.Rotate(angle, 1, 0, 0);
};

Physical.prototype.PitchDown = function(elapsed) {
  const angle = elapsed * this.rotationSpeed[0];
  this.Rotate(-angle, 1, 0, 0);
};

Physical.prototype.RollLeft = function(elapsed) {
  const angle = elapsed * this.rotationSpeed[2];
  this.Rotate(angle, 0, 0, 1);
};

Physical.prototype.RollRight = function(elapsed) {
  const angle = elapsed * this.rotationSpeed[2];
  this.Rotate(-angle, 0, 0, 1);
};

Physical.prototype.Turn = function(angle) {
  // if you aren't rotating about an angle, then you aren't rotating
  if (angle == 0) {
    return;
  }

  const q = new Quaternion();
  q.FromAxisAndAngle(0, 1, 0, angle);
  this.SetOrientation(q.Multiply(this.GetOrientation()));
};

Physical.prototype.TurnLeft = function(elapsed) {
  const angle = elapsed * this.rotationSpeed[1];
  this.Turn(angle);
};

Physical.prototype.TurnRight = function(elapsed) {
  const angle = elapsed * this.rotationSpeed[1];
  this.Turn(-angle);
};

// -------------------------------------
// ------------ POSITION ---------------
// -------------------------------------

/*
 * send as x,y,z
 */
Physical.prototype.SetPosition = function(...args) {
  if (Number.isNaN(this.position[0])) {
    throw new Error('Position became NaN.');
  }
  this.position.Set.apply(this.position, ...args);
  this.modelDirty = true;
};

/*
 * return as Vector
 */
Physical.prototype.GetPosition = function() {
  return this.position;
};

Physical.prototype.ChangePosition = function(...args) {
  if (Number.isNaN(this.position[0])) {
    throw new Error('Position became NaN!');
  }
  this.position.Add.apply(this.position, ...args);
  this.modelDirty = true;
};

// ------------------------------------------
// -----------  MOVEMENT --------------------
// ------------------------------------------
// movement is relative to the physical

/*
 * convertes the local x,y,z vector to the global position vector
 */
Physical.prototype.Warp = function(...args) {
  let x;
  let y;
  let z;
  if (args.length > 1) {
    x = args[0];
    y = args[1];
    z = args[2];
  } else {
    x = args[0][0];
    y = args[0][1];
    z = args[0][2];
  }
  if (x == 0 && y == 0 && z == 0) {
    return;
  }

  // Quaternions don't work correctly if they aren't normalized
  this.orientation.Normalize();

  // get our new position; if we started at 0,0,0
  const d = this.orientation.RotatedVector(x, y, z);

  // add it to our current position to get our new position
  // console.log("Warping vec" + d);
  this.ChangePosition(d);
};

// these movement commands MOVE the physical
// the physical's position is updated in the call
// use the Move commands for player-commanded movement
Physical.prototype.WarpForward = function(distance) {
  this.Warp(0, 0, -distance);
};

Physical.prototype.WarpBackward = function(distance) {
  this.Warp(0, 0, distance);
};

Physical.prototype.WarpLeft = function(distance) {
  this.Warp(-distance, 0, 0);
};

Physical.prototype.WarpRight = function(distance) {
  this.Warp(distance, 0, 0);
};

Physical.prototype.WarpUp = function(distance) {
  this.Warp(0, distance, 0);
};

Physical.prototype.WarpDown = function(distance) {
  this.Warp(0, -distance, 0);
};

// ------------------------------------------
// -----------  VELOCITY --------------------
// ------------------------------------------

// speed is in units per second
Physical.prototype.SetSpeeds = function(...args) {
  this.speed.Set.apply(this.speed, ...args);
};

Physical.prototype.GetSpeeds = function() {
  return this.speed;
};

Physical.prototype.SetSpeed = function(speed) {
  return this.SetSpeeds(speed, speed, speed);
};

Physical.prototype.SetVelocity = function(...args) {
  this.velocity.Set.apply(this.velocity, ...args);
};

Physical.prototype.GetVelocity = function() {
  return this.velocity;
};

Physical.prototype.AddVelocity = function(...args) {
  this.velocity.Add.apply(this.velocity, ...args);
  this.modelDirty = true;
};

// Move commands adjust the velocity
// using the set speed
Physical.prototype.MoveForward = function(elapsed) {
  const distance = elapsed * this.speed[2];
  this.AddVelocity(0, 0, -distance);
};

Physical.prototype.MoveBackward = function(elapsed) {
  const distance = elapsed * this.speed[2];
  this.AddVelocity(0, 0, distance);
};

Physical.prototype.MoveLeft = function(elapsed) {
  const distance = elapsed * this.speed[0];
  this.AddVelocity(-distance, 0, 0);
};

Physical.prototype.MoveRight = function(elapsed) {
  const distance = elapsed * this.speed[0];
  this.AddVelocity(distance, 0, 0);
};

Physical.prototype.MoveUp = function(elapsed) {
  const distance = elapsed * this.speed[1];
  this.AddVelocity(0, distance, 0);
};

Physical.prototype.MoveDown = function(elapsed) {
  const distance = elapsed * this.speed[1];
  this.AddVelocity(0, -distance, 0);
};

// calculates our new position using our current velocity
// and then resets the velocity
Physical.prototype.ApplyVelocity = function() {
  this.Warp(this.velocity);
  this.velocity.Set(0, 0, 0);
};

// ------------------------------------------
// --------------  PARENTING ----------------
// ------------------------------------------

// in order to be a good lineage:
// a camera must be reached
// // therefore it must not infinitely loop
Physical.prototype.IsGoodLineageFor = function(prospectiveChild) {
  const parent = this.GetParent();

  // no parent = no lineage
  if (!parent) {
    return false;
  } else if (parent == prospectiveChild) {
    // the initator already has this physical as an ancestor
    // setting this as a parent would make an infinite loop
    return false;
    // note that we don't check self == prospectiveChild
    // that would throw an error if you tried to reparent to the same parent
    // it's assumed that if its a parent now, its still a good parent;
  }

  return parent.IsGoodLineageFor(prospectiveChild);
};

Physical.prototype.SetParent = function(parent) {
  if (!parent) {
    throw new Error(
        'A Physical must have a parent. Set it to the camera for a default',
    );
  }

  if (!parent.IsGoodLineageFor(this)) {
    throw new Error(
        'Setting this is a parent would result in a lineage that never reaches the camera',
    );
  }
  this.parent = parent;
};

Physical.prototype.GetParent = function() {
  return this.parent;
};

// ------------------------------------------
// -----------  MODELVIEW MATRIX ------------
// ------------------------------------------

Physical.prototype.SetScale = function(...args) {
  this.scale.Set.apply(this.scale, ...args);
  this.modelDirty = true;
};

Physical.prototype.GetScale = function() {
  return this.scale;
};

// combine our position and orientation into a matrix;
Physical.prototype.GetModelMatrix = function() {
  let x;
  let y;
  let z;
  x = this.velocity[0];
  y = this.velocity[1];
  z = this.velocity[2];
  if (x != 0 || y != 0 || z != 0) {
    this.ApplyVelocity();
  }

  // if w == 1 then a 4d vector is a position
  // if w == 0 then a 4d vector is a direction
  if (this.modelDirty) {
    // this.modelMatrix = rotation * translation;
    // this.modelMatrix.FromQuaternionAtVector(self.orientation, self.position);
    const m = this.modelMatrix;
    // this.modelMatrix = rotate * translate * identity
    m.Identity();

    switch (this.modelMode) {
      case PHYSICAL_TRANSLATE_ROTATE_SCALE:
        m.Translate(this.position);
        m.Rotate(this.orientation);
        m.Scale(this.scale);
        break;
      case PHYSICAL_SCALE_ROTATE_TRANSLATE:
        m.Scale(this.scale);
        m.Rotate(this.orientation);
        m.Translate(this.position);
        break;
      case PHYSICAL_ROTATE_TRANSLATE_SCALE:
        m.Rotate(this.orientation);
        m.Translate(this.position);
        m.Scale(this.scale);
        break;
      default:
        throw new Error(
            'Model mode must be an expected value: ' + this.modelMode,
        );
    }

    this.modelDirty = false;
  }

  return this.modelMatrix;
};

// when fully returned it looks like this
// A -> B -> CAM -> A -> B
// Mults as A * B * (CAM * A * B):Inverse()

// in order for this to work properly make sure that the camera's
// GetViewMatrix() is called before any physicals
// otherwise the physical's will be outdated
// this is to prevent having to retrace the camera's lineage more than once

// this would be a good thing to do for any matrix that has many descendants
// say a ship with lots of npcs / players on it

// but this would require a proper ordering of physicals
// which isn't feasible atm
// physicals would need to know who is the child of who.
// something like:
/*
        camera:CalculateViewMatrix();
        while SomePhysicalsNotCalculated do
                for all physicalsNotCalculated do
                        if parentPhysicalCalculated then
                                physical:CalculateViewMatrix()
                        end
                end
        end
*/

// a more feasible method would be to
// to have a bunch of children known by the physical
// then we simply chain down the list starting at the camera;
// this is a far better solution.
/*
        function CalculateViewMatrices()
                self:CalculateViewMatrix(); -- for myself
                for each child in children do
                        child:CalculateViewMatrices() -- for my children
                end
        end
*/
// it starts with a simple camera:CalculateViewMatrices();
// I will return to this.

Physical.prototype.GetViewMatrix = function(...args) {
  // if this was just called then we need to set who sent it
  let requestor;
  if (args.length == 0) {
    requestor = this;
  } else {
    requestor = args[0];
  }

  if (this.parent && this.parent != requestor) {
    this.viewMatrix = this.GetModelMatrix().Multiplied(
        this.parent.GetViewMatrix(requestor),
    );
    return this.viewMatrix;
  } else {
    return this.GetModelMatrix().Inverse();
  }
};

Physical.prototype.GetWorldPositionByViewMatrix = function() {
  return new RMatrix4([
    1,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    1,
    0,
    this.position[0],
    this.position[1],
    this.position[2],
    1,
  ]).Multiply(this.GetViewMatrix().Inverse());
};

// legacy code; left in case I try this again
// it does not work correctly, in all cases
Physical.prototype.GetWorldPosition = function(requestor) {
  const parent = this.parent;
  if (parent && parent != requestor) {
    const rot = parent.GetWorldOrientation(requestor);
    const pos = rot.RotatedVector(this.position);
    return pos.Add(parent.GetWorldPosition(requestor));
  }
  return this.position;
};

// legacy code; left in case I try this again
// it DOES work
Physical.prototype.GetWorldOrientation = function(requestor) {
  const parent = this.parent;
  if (parent && parent != requestor) {
    return parent.GetWorldOrientation(requestor).Multiplied(this.orientation);
  }
  return self.orientation;
};

File: src/alpha/Physical.c
#include "Physical.h"
#include "Cam.h"
#include "Maths.h"
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "../graph/log.h"

struct alpha_Physical* alpha_Physical_new(apr_pool_t* pool, alpha_PhysicalType parentType, void* parent)
{
    struct alpha_Physical* phys;
    if(pool) {
        phys = apr_palloc(pool, sizeof(*phys));
    }
    else {
        phys = malloc(sizeof(*phys));
    }
    phys->pool = pool;

    phys->modelMode = alpha_PHYSICAL_TRANSLATE_ROTATE_SCALE;
    phys->orientation = alpha_Quaternion_new(pool);
    phys->position = alpha_Vector_new(pool);
    phys->modelMatrix = alpha_RMatrix4_new(pool);
    phys->viewMatrix = alpha_RMatrix4_new(pool);
    phys->modelDirty = 0;
    phys->velocity = alpha_Vector_new(pool);
    phys->rotationSpeed = alpha_Vector_create(pool, 1, 1, 1);
    phys->speed = alpha_Vector_create(pool, 5, 5, 5);
    phys->scale = alpha_Vector_create(pool, 1, 1, 1);
    alpha_Physical_SetParent(phys, parentType, parent);
    return phys;
}

void alpha_Physical_destroy(struct alpha_Physical* phys)
{
    alpha_Quaternion_destroy(phys->pool, phys->orientation);
    alpha_Vector_destroy(phys->pool, phys->position);
    alpha_RMatrix4_destroy(phys->pool, phys->modelMatrix);
    alpha_RMatrix4_destroy(phys->pool, phys->viewMatrix);
    alpha_Vector_destroy(phys->pool, phys->velocity);
    alpha_Vector_destroy(phys->pool, phys->rotationSpeed);
    alpha_Vector_destroy(phys->pool, phys->speed);
    alpha_Vector_destroy(phys->pool, phys->scale);
    if(phys->pool) {
        return;
    }
    free(phys);
}

void alpha_Physical_SetOrientation(struct alpha_Physical* phys, float angle, float x, float y, float z)
{
    alpha_Quaternion_Set(phys->orientation, angle, x, y, z);
    phys->modelDirty = 1;

    float q[4];
    alpha_Quaternion_SetIdentity(q);
    alpha_Quaternion_FromAxisAndAngleEach(q, x, y, z, angle);
    alpha_Physical_CopyOrientation(phys, q);
}

void alpha_Physical_CopyOrientation(struct alpha_Physical* phys, float* orientation)
{
    alpha_Quaternion_Copy(phys->orientation, orientation);
    phys->modelDirty = 1;
}

float* alpha_Physical_GetOrientation(struct alpha_Physical* phys)
{
    return phys->orientation;
}

void alpha_Physical_SetRotationSpeeds(struct alpha_Physical* phys, float x, float y, float z)
{
    alpha_Vector_Set(phys->rotationSpeed, x, y, z);
}

float* alpha_Physical_GetRotationSpeeds(struct alpha_Physical* phys)
{
    return phys->rotationSpeed;
}

void alpha_Physical_Rotate(struct alpha_Physical* phys, float angle, float x, float y, float z)
{
    // if you aren't rotating about an angle, then you aren't rotating
    if(angle == 0) {
        return;
    }
    float* q = alpha_QuaternionFromAxisAndAngleEach(0, x, y, z, angle);
    alpha_Quaternion_Multiply(phys->orientation, q);
    alpha_Quaternion_destroy(0, q);
    phys->modelDirty = 1;
}

void alpha_Physical_RotateGlobal(struct alpha_Physical* phys, float angle, float x, float y, float z)
{
    // if you aren't rotating about an angle, then you aren't rotating
    if(angle == 0) {
        return;
    }
    float* q = alpha_QuaternionFromAxisAndAngleEach(0, x, y, z, angle);
    alpha_Quaternion_Multiply(q, phys->orientation);
    alpha_Quaternion_Copy(phys->orientation, q);
    alpha_Quaternion_destroy(0, q);
    phys->modelDirty = 1;
}

void alpha_Physical_YawLeft(struct alpha_Physical* phys, float elapsed)
{
    float angle = elapsed * phys->rotationSpeed[1];
    alpha_Physical_Rotate(phys, angle, 0, 1, 0);
}

void alpha_Physical_YawRight(struct alpha_Physical* phys, float elapsed)
{
    float angle = elapsed * phys->rotationSpeed[1];
    alpha_Physical_Rotate(phys, -angle, 0, 1, 0);
}

void alpha_Physical_PitchUp(struct alpha_Physical* phys, float elapsed)
{
    float angle = elapsed * phys->rotationSpeed[0];
    alpha_Physical_Rotate(phys, angle, 1, 0, 0);
}

void alpha_Physical_PitchDown(struct alpha_Physical* phys, float elapsed)
{
    float angle = elapsed * phys->rotationSpeed[0];
    alpha_Physical_Rotate(phys, -angle, 1, 0, 0);
}

void alpha_Physical_RollLeft(struct alpha_Physical* phys, float elapsed)
{
    float angle = elapsed * phys->rotationSpeed[2];
    alpha_Physical_Rotate(phys, angle, 0, 0, 1);
}

void alpha_Physical_RollRight(struct alpha_Physical* phys, float elapsed)
{
    float angle = elapsed * phys->rotationSpeed[2];
    alpha_Physical_Rotate(phys, -angle, 0, 0, 1);
}

void alpha_Physical_SetPosition(struct alpha_Physical* phys, float x, float y, float z)
{
    alpha_Vector_Set(phys->position, x, y, z);
    phys->modelDirty = 1;
    if(isnan(phys->position[0])) {
        parsegraph_log("Position became NaN.");
    }
}

void alpha_Physical_CopyPosition(struct alpha_Physical* phys, float* src)
{
    alpha_Vector_Copy(phys->position, src);
    phys->modelDirty = 1;
    if(isnan(phys->position[0])) {
        parsegraph_log("Position became NaN.");
    }
}

float* alpha_Physical_GetPosition(struct alpha_Physical* phys)
{
    return phys->position;
}

void alpha_Physical_ChangeEachPosition(struct alpha_Physical* phys, float x, float y, float z)
{
    alpha_Vector_AddEach(phys->position, x, y, z);
    phys->modelDirty = 1;
    if(isnan(phys->position[0])) {
        parsegraph_log("Position became NaN!");
    }
}

void alpha_Physical_ChangePosition(struct alpha_Physical* phys, float* val)
{
    //parsegraph_log("Changed position from (%f, %f, %f) ", phys->position[0], phys->position[1], phys->position[2]);
    alpha_Vector_Add(phys->position, val);
    //parsegraph_log("Changed position to (%f, %f, %f)\n", phys->position[0], phys->position[1], phys->position[2]);
    phys->modelDirty = 1;
    if(isnan(phys->position[0])) {
        parsegraph_log("Position became NaN!\n");
    }
}

void alpha_Physical_Warp(struct alpha_Physical* phys, float* v)
{
    alpha_Physical_WarpEach(phys, v[0], v[1], v[2]);
}

void alpha_Physical_WarpEach(struct alpha_Physical* phys, float x, float y, float z)
{
    if(x == 0 && y == 0 && z == 0) {
        return;
    }

    // Quaternions don't work correctly if they aren't normalized
    //alpha_Quaternion_Normalize(phys->orientation);

    // get our new position; if we started at 0,0,0
    float* d = alpha_Quaternion_RotatedVectorEach(phys->orientation, phys->pool, x, y, z);

    // add it to our current position to get our new position
    //parsegraph_log("Warping vec (%f, %f, %f)\n", d[0], d[1], d[2]);
    alpha_Physical_ChangePosition(phys, d);
}

void alpha_Physical_WarpForward(struct alpha_Physical* phys, float distance)
{
    alpha_Physical_WarpEach(phys, 0, 0, -distance);
};

void alpha_Physical_WarpBackward(struct alpha_Physical* phys, float distance)
{
    alpha_Physical_WarpEach(phys, 0, 0, distance);
}

void alpha_Physical_WarpLeft(struct alpha_Physical* phys, float distance)
{
    alpha_Physical_WarpEach(phys, -distance, 0, 0);
};

void alpha_Physical_WarpRight(struct alpha_Physical* phys, float distance)
{
    alpha_Physical_WarpEach(phys, distance, 0, 0);
};

void alpha_Physical_WarpUp(struct alpha_Physical* phys, float distance)
{
    alpha_Physical_WarpEach(phys, 0, distance, 0);
}

void alpha_Physical_WarpDown(struct alpha_Physical* phys, float distance)
{
    alpha_Physical_WarpEach(phys, 0, -distance, 0);
}

// speed is in units per second
void alpha_Physical_SetSpeeds(alpha_Physical* phys, float x, float y, float z)
{
    alpha_Vector_Set(phys->speed, x, y, z);
}

float* alpha_Physical_GetSpeeds(alpha_Physical* phys)
{
    return phys->speed;
}

/**
 * sets x,y,z and z speeds to the same thing
 */
void alpha_Physical_SetSpeed(alpha_Physical* phys, float speed)
{
    alpha_Physical_SetSpeeds(phys, speed, speed, speed);
}

/**
 * set as x,y,z
 */
void alpha_Physical_SetVelocity(alpha_Physical* phys, float x, float y, float z)
{
    alpha_Vector_Set(phys->velocity, x, y, z);
}

/**
 * return as vector
 */
float* alpha_Physical_GetVelocity(alpha_Physical* phys)
{
    return phys->velocity;
}

void alpha_Physical_AddVelocity(alpha_Physical* phys, float x, float y, float z)
{
    alpha_Vector_AddEach(phys->velocity, x, y, z);
    phys->modelDirty = 1;
}

// Move commands adjust the velocity
// using the set speed
void alpha_Physical_MoveForward(alpha_Physical* phys, float elapsed)
{
    float distance = elapsed * phys->speed[2];
    alpha_Physical_AddVelocity(phys, 0, 0, -distance);
};

void alpha_Physical_MoveBackward(alpha_Physical* phys, float elapsed)
{
    float distance = elapsed * phys->speed[2];
    alpha_Physical_AddVelocity(phys, 0, 0, distance);
};

void alpha_Physical_MoveLeft(alpha_Physical* phys, float elapsed)
{
    float distance = elapsed * phys->speed[0];
    alpha_Physical_AddVelocity(phys, -distance, 0, 0);
};

void alpha_Physical_MoveRight(alpha_Physical* phys, float elapsed)
{
    float distance = elapsed * phys->speed[0];
    alpha_Physical_AddVelocity(phys, distance, 0, 0);
};

void alpha_Physical_MoveUp(alpha_Physical* phys, float elapsed)
{
    float distance = elapsed * phys->speed[1];
    alpha_Physical_AddVelocity(phys, 0, -distance, 0);
};

void alpha_Physical_MoveDown(alpha_Physical* phys, float elapsed)
{
    float distance = elapsed * phys->speed[1];
    alpha_Physical_AddVelocity(phys, 0, distance, 0);
};

// calculates our new position using our current velocity
// and then resets the velocity
void alpha_Physical_ApplyVelocity(alpha_Physical* phys)
{
    alpha_Physical_Warp(phys, phys->velocity);
    alpha_Vector_Set(phys->velocity, 0, 0, 0);
};

//------------------------------------------
//--------------  PARENTING ----------------
//------------------------------------------

// in order to be a good lineage:
// a camera must be reached
// // therefore it must not infinitely loop
int alpha_Physical_IsGoodLineageFor(alpha_Physical* phys, alpha_Physical* prospectiveChild)
{
    void* parent = alpha_Physical_GetParent(phys);

    // no parent = no lineage
    if(!parent) {
        return 0;
    }
    else if(parent == prospectiveChild) {
        // the initator already has this physical as an ancestor
        // setting this as a parent would make an infinite loop
        return 0;
        // note that we don't check self == prospectiveChild
        // that would throw an error if you tried to reparent to the same parent
        // it's assumed that if its a parent now, its still a good parent;
    }

    if(phys->parentType == alpha_PhysicalType_PHYSICAL) {
        return alpha_Physical_IsGoodLineageFor(
            (alpha_Physical*)parent,
            prospectiveChild
        );
    }

    // CAMERAS MAKE THE BEST PARENTS
    return 1;
};

void alpha_Physical_SetParent(alpha_Physical* phys, alpha_PhysicalType parentType, void* parent)
{
    if(!parent) {
        parsegraph_log("A Physical must have a parent. Set it to the camera for a default");
        exit(-1);
        return;
    }

    if(parent && parentType == alpha_PhysicalType_PHYSICAL && !alpha_Physical_IsGoodLineageFor(parent, phys)) {
        parsegraph_log("Setting this is a parent would result in a lineage that never reaches the camera" );
    }
    phys->parentType = parentType;
    phys->parent = parent;
}

void* alpha_Physical_GetParent(alpha_Physical* phys)
{
    return phys->parent;
}

int alpha_Physical_GetParentIsPhysical(alpha_Physical* phys)
{
    return alpha_Physical_GetParentType(phys) == alpha_PhysicalType_PHYSICAL;
}

alpha_PhysicalType alpha_Physical_GetParentType(alpha_Physical* phys)
{
    return phys->parentType;
}

//------------------------------------------
//-----------  MODELVIEW MATRIX ------------
//------------------------------------------

void alpha_Physical_SetScale(struct alpha_Physical* phys, float x, float y, float z)
{
    alpha_Vector_Set(phys->scale, x, y, z);
    phys->modelDirty = 1;
}

float* alpha_Physical_GetScale(struct alpha_Physical* phys)
{
    return phys->scale;
}

// combine our position and orientation into a matrix;
float* alpha_Physical_GetModelMatrix(struct alpha_Physical* phys)
{
    float x, y, z;
    x = phys->velocity[0];
    y = phys->velocity[1];
    z = phys->velocity[2];
    if(x != 0 || y != 0 || z != 0) {
        alpha_Physical_ApplyVelocity(phys);
    }

    // if w == 1 then a 4d vector is a position
    // if w == 0 then a 4d vector is a direction
    if(phys->modelDirty) {
        // this.modelMatrix = rotation * translation;
        // this.modelMatrix.FromQuaternionAtVector(self.orientation, self.position);
        float* m = phys->modelMatrix;
        // this.modelMatrix = rotate * translate * identity
        alpha_RMatrix4_SetIdentity(m);

        switch(phys->modelMode) {
        case alpha_PHYSICAL_TRANSLATE_ROTATE_SCALE:
            alpha_RMatrix4_Translate(m, phys->position);
            alpha_RMatrix4_Rotate(m, phys->orientation);
            alpha_RMatrix4_Scale(m, phys->scale);
            break;
        case alpha_PHYSICAL_SCALE_ROTATE_TRANSLATE:
            alpha_RMatrix4_Scale(m, phys->scale);
            alpha_RMatrix4_Rotate(m, phys->orientation);
            alpha_RMatrix4_Translate(m, phys->position);
            break;
        case alpha_PHYSICAL_ROTATE_TRANSLATE_SCALE:
            alpha_RMatrix4_Rotate(m, phys->orientation);
            alpha_RMatrix4_Translate(m, phys->position);
            alpha_RMatrix4_Scale(m, phys->scale);
            break;
        }

        phys->modelDirty = 0;
    }

    return phys->modelMatrix;
}

// when fully returned it looks like this
// A -> B -> CAM -> A -> B
// Mults as A * B * (CAM * A * B):Inverse()

// in order for this to work properly make sure that the camera's
// GetViewMatrix() is called before any physicals
// otherwise the physical's will be outdated
// this is to prevent having to retrace the camera's lineage more than once

// this would be a good thing to do for any matrix that has many descendants
// say a ship with lots of npcs / players on it

// but this would require a proper ordering of physicals
// which isn't feasible atm
// physicals would need to know who is the child of who.
// something like:
/*
	camera:CalculateViewMatrix();
	while SomePhysicalsNotCalculated do
		for all physicalsNotCalculated do
			if parentPhysicalCalculated then
				physical:CalculateViewMatrix()
			end
		end
	end	
*/

// a more feasible method would be to
// to have a bunch of children known by the physical
// then we simply chain down the list starting at the camera;
// this is a far better solution.
/*
	function CalculateViewMatrices()
		self:CalculateViewMatrix(); -- for myself
		for each child in children do
			child:CalculateViewMatrices() -- for my children
		end
	end
*/
// it starts with a simple camera:CalculateViewMatrices();
// I will return to this.

float* alpha_Physical_GetViewMatrix(alpha_Physical* phys, void* requestor)
{
    // if this was just called then we need to set who sent it
    if(!requestor) {
        requestor = phys;
    }

    if(phys->parent && phys->parent != requestor) {
        if(phys->parentType == alpha_PhysicalType_PHYSICAL) {
            phys->viewMatrix = alpha_RMatrix4_Multiplied(phys->pool,
                alpha_Physical_GetModelMatrix(phys),
                alpha_Physical_GetViewMatrix(phys->parent, requestor)
            );
        }
        else {
            // It's a camera.
            phys->viewMatrix = alpha_RMatrix4_Multiplied(phys->pool,
                alpha_Physical_GetModelMatrix(phys),
                alpha_Camera_GetViewMatrix(phys->parent, requestor)
            );
        }
        return phys->viewMatrix;
    }

    float* modelMatrix = alpha_Physical_GetModelMatrix(phys);
    return alpha_RMatrix4_Inversed(phys->pool, modelMatrix);
}

float* alpha_Physical_GetWorldPositionByViewMatrix(alpha_Physical* phys)
{
    float* rv = alpha_RMatrix4_Create(phys->pool,
        1,0,0,0,
        0,1,0,0,
        0,0,1,0,
        phys->position[0], phys->position[1], phys->position[2], 1
    );
    float* vm = alpha_Physical_GetViewMatrix(phys, 0);
    alpha_RMatrix4_Inverse(vm);
    alpha_RMatrix4_Multiply(rv, vm);
    return rv;
}


// legacy code; left in case I try this again
// it does not work correctly, in all cases
float* alpha_Physical_GetWorldPosition(alpha_Physical* phys, void* requestor)
{
    void* parent = phys->parent;
    if(parent && phys->parentType == alpha_PhysicalType_PHYSICAL && parent != requestor) {
        float* rot = alpha_Physical_GetWorldOrientation(parent, requestor);
        float* pos = alpha_Quaternion_RotatedVectorEach(rot, phys->pool, phys->position[0], phys->position[1], phys->position[2]);
        alpha_Vector_Add(pos, alpha_Physical_GetWorldPosition(parent, requestor));
        return pos;
    };
    return phys->position;
}

// legacy code; left in case I try this again
// it DOES work
float* alpha_Physical_GetWorldOrientation(alpha_Physical* phys, void* requestor)
{
    void* parent = phys->parent;
    if(parent && phys->parentType == alpha_PhysicalType_PHYSICAL && parent != requestor) {
        return alpha_Quaternion_Multiplied(
            phys->pool,
            alpha_Physical_GetWorldOrientation(parent, requestor),
            phys->orientation
        );
    }
    return phys->orientation;
}

void alpha_Physical_Dump(struct alpha_Physical* phys)
{
    fprintf(stderr, "([%f, %f, %f], {%f, %f, %f, %f})\n",
        phys->position[0],
        phys->position[1],
        phys->position[2],
        phys->orientation[0],
        phys->orientation[1],
        phys->orientation[2],
        phys->orientation[3]
    );
}
